webhook/mutator.go

```diff
@@ -125,55 +125,79 @@ func unMarshall(req *v1beta1.AdmissionRequest) (corev1.Pod, error) {
 	return pod, err
 }
 
-func shouldMutate(ignoredList []string, metadata *metav1.ObjectMeta) bool {
+func shouldMutate(ignoredList []string, metadata *metav1.ObjectMeta) ([]string, bool) {
 	for _, namespace := range ignoredList {
 		if metadata.Namespace == namespace {
 			glog.Infof("Skipping mutation for [%v] in special namespace: [%v]", metadata.Name, metadata.Namespace)
-			return false
+			return nil, false
 		}
 	}
 
-	required := false
 	annotations := metadata.GetAnnotations()
 	if annotations == nil {
 		annotations = map[string]string{}
 	}
 
-	status := annotations[sideCarInjectionStatusAnnotation]
-	if strings.ToLower(status) != injectedValue {
-		switch strings.ToLower(annotations[sideCarInjectionAnnotation]) {
-		case "y", "yes", "true", "on":
-			required = true
+	if status, ok := annotations[sideCarInjectionStatusAnnotation]; ok && strings.ToLower(status) == injectedValue {
+		glog.Infof("Skipping mutation for [%v]. Has been mutated already", metadata.Name)
+		return nil, false
+	}
+
+	if sidecars, ok := annotations[sideCarInjectionAnnotation]; ok {
+		parts := strings.Split(sidecars, ",")
+		for i := range parts {
+			parts[i] = strings.Trim(parts[i], " ")
+		}
+
+		if len(parts) > 0 {
+			glog.Infof("sideCar injection for %v/%v: sidecars: %v", metadata.Namespace, metadata.Name, sidecars)
+			return parts, true
 		}
 	}
 
-	glog.Infof("sideCar injection for %v/%v: status: %q required:%v", metadata.Namespace, metadata.Name, status, required)
-	return required
+	glog.Infof("Skipping mutation for [%v]. No action required", metadata.Name)
+	return nil, false
 }
 
-func createPatch(pod *corev1.Pod, inSidecarConfig *SideCar, annotations map[string]string) ([]byte, error) {
-	sideCar, err := makeCopy(inSidecarConfig)
-	if err != nil {
-		return nil, err
-	}
+func createPatch(pod *corev1.Pod, sideCarNames []string, sideCars map[string]*SideCar, annotations map[string]string) ([]byte, error) {
+
+	var patch []patchOperation
+	var containers       []corev1.Container
+	var volumes          []corev1.Volume
+	var imagePullSecrets []corev1.LocalObjectReference
+	count := 0
+
+	for _, name := range sideCarNames {
```

Comment:
Will it only be able to append containers, volumes, imagePullSecrets ?

Reply:
@ayansen that is all we are having in `webhook.Sidecar` struct.

Reply:
Looking at Pod spec, these covers most of what is needed to inject a sidecar

Reply:
sounds good. I was thinking about initContainers but I guess most sidecars wouldn't need one. This looks good for now